scripts/get_nbd_extents.py

```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/python
+
+"""
+Returns a list of extents with their block statuses for an NBD export.
+
+This program uses new NBD capabilities introduced in QEMU 2.12.
+
+It uses the BLOCK_STATUS NBD extension, which is documented here:
+https://github.com/NetworkBlockDevice/nbd/blob/extension-blockstatus/doc/proto.md
+The structured replies functionality that this extension relies on is in the
+main branch of the NBD protocol docs:
+https://github.com/NetworkBlockDevice/nbd/blob/master/doc/proto.md
+"""
+
+import argparse
+import json
+import logging
+import logging.handlers
+
+from python_nbd_client import PythonNbdClient
+import python_nbd_client
+
+
+LOGGER = logging.getLogger("get_nbd_extents")
+LOGGER.setLevel(logging.DEBUG)
+# The log level of python_nbd_client is not set, therefore it will default to
+# that of the root logger, which is WARNING by default.
+
+# Request length is a 32-bit int.
+# It looks like for qemu 2.12, the length parameter of a NBD_CMD_BLOCK_STATUS
+# request is not limited by the maximum block size supported by the server (as
+# defined by NBD_INFO_BLOCK_SIZE), only by the size of a 32-bit int.
+MAX_REQUEST_LEN = 2 ** 32 - 1
+
+# Make the NBD_CMD_BLOCK_STATUS request aligned to 512 bytes, just in case. But
+# it looks like this is not required for qemu 2.12.
+MAX_REQUEST_LEN = MAX_REQUEST_LEN - (MAX_REQUEST_LEN % 512)
+
+def _get_extents(path, exportname):
+    with PythonNbdClient(
+        address=path, exportname=exportname, unix=True, use_tls=False, connect=False) as client:
+
+        client.negotiate_structured_reply()
+
+        # Select our metadata context. This context is documented at
+        # https://github.com/NetworkBlockDevice/nbd/blob/extension-blockstatus/doc/proto.md#baseallocation-metadata-context
+        context = 'base:allocation'
+        selected_contexts = client.set_meta_contexts(exportname, [context])
+        assert len(selected_contexts) == 1
+        (meta_context_id, meta_context_name) = selected_contexts[0]
+        assert meta_context_name == context
+
+        client.connect(exportname)
+
+        size = client.get_size()
+        LOGGER.debug(
+            'Connected to NBD export %s served at path %s of size %d bytes',
+            exportname, path, size)
+
+        offset = 0
+        while offset < size:
+            request_len = min(MAX_REQUEST_LEN, size - offset)
+            replies = client.query_block_status(offset, request_len)
```

Comment:
Those block sizes do not apply to this request in case of QEMU, for some reason. So we can (and we do) send a request that exceeds the maximum block size as advertised by the NBD server via NBD_OPT_INFO. We don't have to advertise aligned reads, reads just have to be aligned to the minimum block size of the server.
QEMU 2.12 serving a QCOW image returned the following block sizes for me:
```
 'maximum_block_size': 33554432 (2^25), 'preferred_block_size': 4096, 'minimum_block_size': 512
```

Reply:
And actually, we do benefit from having such a large `MAX_REQUEST_LEN`, because it seems that the max extent that qemu can return is almost this value, 4294966784, which is weirdly 511 less than  2 ^ 32 - 1:
```
[root@minnow ~]# python /opt/xensource/libexec/get_nbd_extents.py --path /var/run/qemu-nbd.fb918427-4be9-4cc5-b1fe-5ba1ed220f30 --exportname qemu_node
[{"length": 65536, "flags": 0}, {"length": 983040, "flags": 3}, {"length": 16973824, "flags": 0}, {"length": 25100288, "flags": 3}, {"length": 4294966784, "flags": 0}, {"length": 4294966784, "flags": 0}, {"length": 4294966784, "flags": 0}, {"length": 796919296, "flags": 0}, {"length": 1638400, "flags": 3}, {"length": 28901376, "flags": 0}, {"length": 4290314240, "flags": 3}, {"length": 1357316096, "flags": 0}, {"length": 147062784, "flags": 3}, {"length": 62717952, "flags": 0}, {"length": 63045632, "flags": 3}, {"length": 314638336, "flags": 0}, {"length": 4294966784, "flags": 3}, {"length": 1483145728, "flags": 3}, {"length": 65536, "flags": 0}, {"length": 1048576, "flags": 3}]
```